### PR TITLE
Adds Read Access granted event

### PR DESCRIPTION
**Pull request by DorianWilhelm**
- emits event that read access has been granted with `reader`, `tokenId` and `until`
- updates tests

### CODE REVIEW

File: src/IPNFT.sol [hunk 1]
```diff
@@ -63,6 +63,7 @@ contract IPNFT is
 
     event Reserved(address indexed reserver, uint256 indexed reservationId);
     event IPNFTMinted(address indexed owner, uint256 indexed tokenId, string tokenURI);
+    event ReadAccessGranted(address indexed reader, uint256 indexed tokenId, uint256 until);
```
Comment (elmariachi111) [inline src/IPNFT.sol:66]: I would put the tokenId in the first topic here. Contextually it's slightly different than the other ones which read ("owner created NFT tokenId with uri"). Yours reads ("tokenId has granted read access to reader")

Comment (DorianWilhelm) [reply src/IPNFT.sol:66]: good point!

**Review by elmariachi111 — APPROVED** [review]
lgtm

**Comment by elmariachi111** [comment]
adds subgraph support
image

**Comment by elmariachi111** [comment]
better "ids" using eth's padding for hashes (see inline comment):
a "canRead" entity is id'd by `keccak256(abi.encode(tokenId,address))`. 


image

**Comment by firattale** [comment]
lgtm from my side, I know this is not related to this PR, but I have a suggestion on the general function accessors on IP-NFT Contracts.

I think we can use `external` instead of `public` for all functions because they are not internally called and also external is cheaper to use (public funcs writes all of the arguments to memory). Also it is considered one of the best practices. We can discuss and implement it on another ticket. wdyt? @DorianWilhelm @elmariachi111

**Comment by elmariachi111** [comment]
what do you mean with "considered best practice"? It's a technical difference if a method is just public or becomes part of an extractable interface by being marked as external. But imo I agree that functions that *would* become part of a public interface should be marked as external. A good counterexample are the changes from the symbol PR that need to do internal function calls and imo it's really sth different if you call a function by its name or by prefixing it with `this` (which changes msg.sender). So, I'd say, lets change that on the final merge.

**Comment by firattale** [comment]
> what do you mean with "considered best practice"? It's a technical difference if a method is just public or becomes part of an extractable interface by being marked as external. But imo I agree that functions that _would_ become part of a public interface should be marked as external. A good counterexample are the changes from the symbol PR that need to do internal function calls and imo it's really sth different if you call a function by its name or by prefixing it with `this` (which changes msg.sender). So, I'd say, lets change that on the final merge.

I mean there are some unofficial standards that we can also follow like using correct function accessors. I think if we are sure that these funcs are not going to called internally, it would be safe and cheap to use `external`, we can implement it on a later merge as you said :)

**Review by firattale — APPROVED** [review]

**Pull request merged**